### PR TITLE
Adding support for equi-join on struct

### LINKED ISSUE #7543: [FEA] Add equi-join support for structs
There is no current way to run a join using a struct column. It would be nice to support this in cudf. This will probably leverage some of the work from #7422.

**Describe the solution you'd like**
It would certainly be nice to support arbitrary depths, but based on #7422 it might require a first-pass at a single depth and a later PR to add support for more depths.

Of particular concern is handling of NaN, Inf, -Inf, etc. These should be handled exactly the same as if the comparison was made between columns of the types inside the structs, so this should be fairly straight forward. The other concern is null handling. Join has the option of nulls comparing equal or not based on incoming parameters. There is some potential for issues around nulls here and care needs to be taken with the implementation.

A null struct would not be the same as a struct with all null entries.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
Adds support for equijoin on structs.

This PR is leveraging the [struct PR](https://github.com/rapidsai/cudf/pull/7422) and the [rewrite for join API](https://github.com/rapidsai/cudf/pull/7454). It enables equijoin on structs by flattening the struct for the hash calculation.

closes #7543

### CODE REVIEW

**Comment by devavret** [comment]
Ok if I review this after the noise dies down and all the dependent PRs get merged?

**Comment by hyperbolic2346** [comment]
> Ok if I review this after the noise dies down and all the dependent PRs get merged?

Certainly and I would hope you would as the noise on this is extreme due to the branches involved. I'm waiting for the dependencies to get merged and I'll rebase on branch-0.19 and it should be much easier to reason about then.

**Comment by hyperbolic2346** [comment]
Just noticed that I didn't call out the dependency. it is https://github.com/rapidsai/cudf/pull/7454

**Comment by hyperbolic2346** [comment]
This is now ready for review.

**Comment by harrism** [comment]
> This PR is leveraging struct PR and rewrite for join API. 

Not sure what you mean by this. Please clarify with readers of the changelog in mind, since this will get pasted into it.

**Review by nvdbaranec — COMMENTED** [review]
Surprisingly simple.   Does `flatten_nested_columns` throw an exception if it encounters a list column?

**Comment by hyperbolic2346** [comment]
> Surprisingly simple. Does `flatten_nested_columns` throw an exception if it encounters a list column?

It marches the table columns looking for structs. Doesn’t complain about lists.

`if (child.type().id() == type_id::STRUCT)`

**Comment by nvdbaranec** [comment]
> 
> 
> > Surprisingly simple. Does `flatten_nested_columns` throw an exception if it encounters a list column?
> 
> It marches the table columns looking for structs. Doesn’t complain about lists.
> 
> `if (child.type().id() == type_id::STRUCT)`

So if you had this:

`struct<int, struct<list<struct<...>> , list, struct<int>>>>`

it would flatten to

`int, list, list, int` ?  That second list quietly contains a struct within it.  Does that mess things up?  Or is it the case that the mere presence of list columns in any case will cause the join to throw later on?

File: cpp/src/join/join.cu [hunk 1]
```diff
@@ -34,10 +35,15 @@ inner_join(table_view const& left_input,
            rmm::cuda_stream_view stream,
            rmm::mr::device_memory_resource* mr)
 {
+  // flatten any structs out. Note this happens before dictionary matching because
+  // structs can contain dictionaries.
+  auto const flattened_left  = structs::detail::flatten_nested_columns(left_input, {}, {});
+  auto const flattened_right = structs::detail::flatten_nested_columns(right_input, {}, {});
```
Comment (jrhemstad) [inline cpp/src/join/join.cu:41]: I think this should just happen in the `hash_join` ctor and `probe`. That way these calls don't need to be repeated.

**Comment by karthikeyann** [comment]
@nvdbaranec I am working on templatized code for `flatten_nested_columns`. (still doesn't have list support, but throws error). Targeted for 0.20 release.

**Comment by hyperbolic2346** [comment]
@nvdbaranec I just coded this up as a quick test and we get "C++ exception with description "cuDF failure at: ../../../../include/cudf/table/row_operators.cuh:363: Attempted to compare elements of uncomparable types." thrown. This is the same error as passing a list by itself to join.

File: cpp/src/join/join.cu (hunk 1 above)
Comment (hyperbolic2346) [reply cpp/src/join/join.cu:41]: I originally had it this way, but I was worried that the dictionary matching wouldn't happen for a dictionary inside a struct. The dictionary code is duplicated as well. Should we move both inside `hash_join` and `probe`?

Comment (jrhemstad) [reply cpp/src/join/join.cu:41]: The dictionary code can't be moved inside the `hash_join` as it requires knowing both of the dictionary columns at the same time.

**Comment by harrism** [comment]
> > Not sure what you mean by this. Please clarify with readers of the changelog in mind, since this will get pasted into it.
> 
> Am I correct in thinking that only the first line is used in the changelog? I have updated the comment to have a one-liner for what the change is on a high level and then detail for reviewers below that.

Nope, not correct. :)

**Review by ttnghia — APPROVED** [review]

File: cpp/src/join/join.cu (hunk 1 above)
Comment (hyperbolic2346) [reply cpp/src/join/join.cu]: Spoke with Jake and we don't need to run the flatten before the dictionary work, so I have moved this code into the `hash_join` object as requested. This reduces the number places for this flattening code to a single class.

**Review by devavret — APPROVED** [review]
This is such a neat little PR.

**Review by jrhemstad — APPROVED** [review]

**Comment by hyperbolic2346** [comment]
@gpucibot merge

**Comment by hyperbolic2346** [comment]
@gpucibot merge

**Pull request merged**